src/main/java/com/google/devtools/build/lib/remote/RemoteLeaseExtension.java

```diff
@@ -206,13 +208,20 @@ private void extendLeaseForAction(
             token.extendOutputFile(artifact, expirationTime);
           }
         }
+      } else {
+        delete = true;
+        break;
       }
     }
 
-    if (actionCache != null && token != null && token.dirty) {
-      // Only update the action cache entry if the token was updated because it usually involves
-      // serialization.
-      actionCache.put(token.key, token.entry);
+    if (actionCache != null && token != null) {
+      if (delete) {
+        actionCache.remove(token.key);
```

Comment:
Can you elaborate why we need to delete entry from the action cache?

The changes in `RemoteOutputChecker` below allows cold build to hit action cache even if the TTL of cache entries expired. Is that the case you want to avoid?